design-documents/service-isolation/caching-layer.md

```diff
@@ -0,0 +1,105 @@
+Also private content cache on client-side should expire according to TTL.
+
+ ![Private content caching](img/private-cache.jpg)
+ 
+ In general, the same caching approach works in current monolith Magento architecture.
+ For service oriented architecture next issues should be solved:
+  * The http-response header should include a set of tags for effective cache purging on reverse-proxy.
+ The set should be created while request is going via the chain of services. Appropriate design should be prepared.
+  * Caching on reverse-proxy should be aligned with request authorization approach. The presence of an authorization token should be taken into account.
+  
+
+### Application data caching
+
+Service application needs to cache such data as DB query results, merged configurations for efficiently and rapidly respond to requests.
+
+The three possible types of caches are the following:
+ * Local caches
+ * Remote caches dedicated to each service
+ * Remote caches shared across the services
+ 
+#### Local caches
+ - Pros: fast, no network traffic associated with retrieving data
+ - Cons: 
+   - shares sources with application
+   - can lead to data inconsistency when multiple instances of application servers are created
+   - the information stored within an individual cache node cannot be shared with other application servers
+   - not follow the 12-factor principles (any data that needs to persist must be stored in a stateful backing service)
+ 
+ ![Local cache](img/local-app-cache.png)
+ 
+#### Remote caches dedicated to each service
+ - Pros: 
+   - scalable, adheres to 12-factor principles
+   - externalize application state
+   - possibility dynamically change the number of service instances
+   - contains only dedicated service data
+ - Cons: 
+   - a slight delay in network calls
+   - difficulties with centralized cache invalidation
+ 
+  ![Remote cache dedicated to each service](img/remote-dedicated-cache.png)
+ 
+#### Remote caches shared across the services
+ - Pros: 
+   - centralized cache invalidation is easy
+   - scalable
+   - adheres to 12-factor principles
+ - Cons: 
+   - cache nodes contain mixed data from different services
+ 
+ ![Remote cache shared across the services](img/remote-cache-cluster.png)
+ 
+Local caches are isolated on the individual nodes and cannot be leveraged within a cluster of application servers.
+Remote distributed caches on the other hand, provide low latency and higher levels of availability, when employed with read replicas and can provide a shared environment for all application servers to utilize the cached data.
+Remote cache dedicated to each service allows to store only data of particular service. 
+But there is no possibility (needs implementation) to flush all cache instances in case of upgrade or other huge changes.
+Remote caches shared across the services are easy to flush centralized, but mix data from different services.
+
+
+#### Cache invalidation
```

Comment:
How can we invalidate cache for all services and BFF? Each service can expose API for cache invalidation, just like you suggest. But better approach would be is to have services and BFF listen to cache invalidation events. This way system would be less coupled.

Reply:
As I remember Anton said that we should go away from service bus in interservice communication as it can be a single point of failure. So I didn't include this way in document